File: Answers-C++/047.cpp
// Project Euler Problem 47
//
// The first two consecutive numbers to have two distinct prime factors are:
// 
// 14 = 2 × 7
// 15 = 3 × 5
// 
// The first three consecutive numbers to have three distinct prime factors are:
// 
// 644 = 2² × 7 × 23
// 645 = 3 × 5 × 43
// 646 = 2 × 17 × 19.
// 
// Find the first four consecutive integers to have four distinct prime factors. 
// What is the first of these numbers?
//
// Answer: 134043

#include <cmath>
#include <iostream>
#include <vector>

namespace
{
    const int TARGET_FACTOR_COUNT = 4;
    const int TARGET_RUN_LENGTH = 4;
    const int MAXIMUM_VALUE = 150000;
}

/*
 * Sieve of Eratosthenes (integers)
 * Simple, ancient algorithm for finding all prime numbers up to any given limit.
 * It does so by iteratively marking as composite (i.e., not prime) the multiples
 * of each prime, starting with the multiples of 2.
 *
 * Provides a vector of integers that represent the prime numbers under the maximum value
*/
void GetPrimes (const int maximum, std::vector<int> &primes)
{
    // Clear the input vector in case there is existing data
    primes.clear();
    
    // There are no primes less than 2
    if (maximum < 2)
        return;
    
    // Construct and execute the Sieve
    const int sqrtMaximum = static_cast<int>( sqrt(static_cast<double>(maximum)) );
    std::vector<bool> primeTracker (maximum, true);
    
    for (int i = 2; i < sqrtMaximum; ++i)
    {
        if (!primeTracker[i])
            continue;
        
        for (int j = i + i; j < maximum; j += i)
        {
            if (j < 0)  // Guard against integer overflow
                break;
            
            primeTracker[j] = false;
        }
    }
    
    // Generate the list of primes to return
    for (int k = 2; k < maximum; ++k)
    {
        if (primeTracker[k] == true)
            primes.push_back (k);
    }
}

/*
 * Get a list of the prime factors of a given integer
 * This function expects to have a valid list of primes passed in
 */
int GetPrimeFactors (const int number, const std::vector<int> primes)
{
    int factorCount = 0;
    int lastFactor = 0;
    int remainder = number;
    
    // Keep looping until we reach the end of our dividing
    while (remainder != 1 )
    {
        // Determine the lowest prime that we can use to divide the remainder
        for (int i = 0; i < primes.size(); i++)
        {
            if (remainder <= 1) // Sanity check
                break;
            
            // Found a prime factor
            if (0 == remainder % primes[i])
            {
                // If the current factor is the same as
                // the last one then don't count it
                if (primes[i] != lastFactor)
                {
                    factorCount++;
                    lastFactor = primes[i];
                }
                
                remainder /= primes[i];
                break;
            }
        }
    }

    return factorCount;
}

// We are using brute force for the prime factor search
// This takes about sixteen seconds on my machine
// There are clearly some ways that we could speed this up

int main (int argc, char *argv[]) 
{
    std::vector<int> primes;
    GetPrimes (MAXIMUM_VALUE, primes);
    
    int leadValue = 0;
    int consecutivePrimeCount = 0;
    
    // Examine all integers up to the maximum value
    // Start at one so we don't try to factor zero
    for (int i = 1; i < MAXIMUM_VALUE; i++)
    {   
        int factorCount = GetPrimeFactors (i, primes);
        
        // Reset the counter if we didn't get the right number of factors
        if (factorCount != TARGET_FACTOR_COUNT)
        {
            consecutivePrimeCount = 0;
            continue;
        }
        
        // If this is the first in a new potential sequence 
        // then keep track of the lead integer value
        if (consecutivePrimeCount == 0)
            leadValue = i;
        
        consecutivePrimeCount++;
        
        // Determine if we have found the target run
        if (consecutivePrimeCount == TARGET_RUN_LENGTH)
            break;
    }
    
    std::cout << "The target integer is: " << leadValue << std::endl;
}

File: Answers-Python/047.py
#!/usr/bin/python

# Project Euler Problem 47
#
# The first two consecutive numbers to have two distinct prime factors are:
# 
# 14 = 2 x 7
# 15 = 3 x 5
# 
# The first three consecutive numbers to have three distinct prime factors are:
# 
# 644 = 2^2 x 7 x 23
# 645 = 3 x 5 x 43
# 646 = 2 x 17 x 19.
# 
# Find the first four consecutive integers to have four distinct prime factors. 
# What is the first of these numbers?
#
# Answer: 134043

import math

TARGET_FACTOR_COUNT = 4
TARGET_RUN_LENGTH = 4
MAXIMUM_VALUE = 150000


# Sieve of Eratosthenes
# Simple, ancient algorithm for finding all prime numbers up to any given limit.
# It does so by iteratively marking as composite (i.e., not prime) the multiples
# of each prime, starting with the multiples of 2.
def get_primes(maximum):
    # There are no primes less than 2
    if maximum < 2:
        return

    # Construct and execute the Sieve
    sqrt_maximum = math.sqrt(maximum)
    primes = []
    prime_tracker = []

    for i in range(maximum):
        prime_tracker.append(True)

    for i in range(2, int(sqrt_maximum)):
        if not prime_tracker[i]:
            continue

        for j in range(i + i, maximum, i):
            prime_tracker[j] = False

    # Generate the list of primes to return
    for k in range(2, maximum):
        if prime_tracker[k]:
            primes.append(k)

    return primes


# Get a list of the prime factors of a given integer
# This function expects to have a valid list of primes passed in
def get_prime_factors(number, primes):
    factor_count = 0
    last_factor = 0
    remainder = number

    # Keep looping until we reach the end of our dividing
    while remainder != 1:
        # Determine the lowest prime that we can use to divide the remainder
        for i in range(0, len(primes)):
            if remainder <= 1:  # Sanity check
                break

            # Found a prime factor
            if 0 == remainder % primes[i]:
                # If the current factor is the same as
                # the last one then don't count it
                if primes[i] != last_factor:
                    factor_count += 1
                    last_factor = primes[i]

                remainder /= primes[i]
                break

    return factor_count


# We are using brute force for the prime factor search
# This takes about sixteen seconds on my machine
# There are clearly some ways that we could speed this up

primes = get_primes(MAXIMUM_VALUE)

lead_value = 0
consecutive_prime_count = 0

# Examine all integers up to the maximum value
# Start at one so we don't try to factor zero
for i in range(1, MAXIMUM_VALUE):
    factor_count = get_prime_factors(i, primes)

    # Reset the counter if we didn't get the right number of factors
    if factor_count != TARGET_FACTOR_COUNT:
        consecutive_prime_count = 0
        continue

    # If this is the first in a new potential sequence
    # then keep track of the lead integer value
    if consecutive_prime_count == 0:
        lead_value = i

    consecutive_prime_count += 1

    # Determine if we have found the target run
    if consecutive_prime_count == TARGET_RUN_LENGTH:
        break

print ("The target integer is: ", lead_value)
